File: python/modprop/core/modules_core.py
"""This module contains base classes and types for creating new Modules and using module trees.
"""
import abc
from collections import deque
import numpy as np

class ModuleBase(object):
    """The base interface for all modules. Modules must inherit from this interface.
    """
    __metaclass__ = abc.ABCMeta

    def __init__(self):
        self._inputs = []
        self._outputs = []

    def register_inputs(self, *args):
        """Registers inputs to this module.

        Parameters
        ----------
        inputs : Variable number of inputs to register.
        """
        for arg in args:
            if not isinstance(arg, InputPort):
                raise ValueError('All inputs must be InputPort type')
            self._inputs.append(arg)

    def register_outputs(self, *args):
        """Registers outputs to this module.

        Parameters
        ----------
        outputs : Variable number of outputs to register.
        """
        for arg in args:
            if not isinstance(arg, OutputPort):
                raise ValueError('All outputs must be OutputPort type')
            self._outputs.append(arg)

    def foreprop_ready(self):
        """Returns if the module is ready to forward-propagate.

        Default implementation returns true when all inputs are ready and
        not all outputs are set.

        Returns
        -------
        ready : Boolean denoting if the module is ready to foreprop
        """
        return all(self._inputs) and not all(self._outputs)

    @abc.abstractmethod
    def foreprop(self):
        """Perform forward-propagation for this module.

        Returns
        -------
        ready : The aggregated return list from all forepropped output ports.
        """
        return []

    def backprop_ready(self):
        """Returns if the module is ready to backward-propagate.

        Typically this is when all outputs have received all backprops.
        Default implementation checks to see if all outputs are ready to backprop.

        Returns
        -------
        ready : Boolean denoting if the module is ready to backprop
        """
        return all([o.backprop_ready() for o in self._outputs])

    @abc.abstractmethod
    def backprop(self):
        """Perform backward-propagation for this module.

        Returns
        -------
        ready : The aggregated return list from all backpropped input ports.
        """
        return []

    def is_invalid(self):
        """Returns if the module is fully invalidated.

        Typically this is when all ports are invalidated.
        Default implementation checks to see if all ports are invalidated.

        Returns
        -------
        invalid : Boolean denoting if this module is fully invalid
        """
        return not any(self._inputs) and not any(self._outputs)

    def invalidate(self):
        """Invalidate this modules' inputs and outputs.

        Default implementation first checks to see if the module is already invalid.
        If it is not, it calls invalidate on all inputs and outputs.

        Returns
        -------
        ready : List of modules to invalidate next.
        """
        if self.is_invalid():
            return []

        ready = []
        for i in self._inputs:
            ready += i.invalidate()
        for o in self._outputs:
            ready += o.invalidate()
        return ready

# TODO Ways to unregister port connections
class InputPort(object):
    """An input to a module. Ideally instantiated as a member of the module.

    Parameters
    ----------
    module : The owning module. Must implement the ModuleBase interface.
    """
    def __init__(self, module):
        if not isinstance(module, ModuleBase):
            raise ValueError('module must implement ModuleBase')

        self._module = module
        self._value = None
        self._source = None

    def __nonzero__(self):
        """Override of Python boolean test operator to return if the port has a value.

        Returns
        -------
        ready : Boolean denoting if the port has a valid value.
        """
        return self._value is not None

    def invalidate(self):
        """Invalidate this input port and propagate to the module and source.

        Returns
        -------
        valid : List of modules to invalidate next.
        """
        # If we're already invalidated, there's nothing for us to do here
        if not self:
            return []

        self._value = None
        valid = []

        # If the owning module is not invalid, return it
        if not self._module.is_invalid():
            valid.append(self._module)

        # Propagate invalidation to source
        if self._source is not None:
            valid += self._source.invalidate()

        return valid

    def foreprop(self, v):
        """Set this port's value and forward-propagate.

        Typically only called by OutputPorts.

        Parameters
        ----------
        v : The value to set the port to.

        Returns
        -------
        ready : List of modules to foreprop next.
        """
        self._value = v

        if self._module.foreprop_ready():
            return [self._module]
        else:
            return []

    def backprop(self, do_dx):
        """Give this port a backpropagation accumulator to pass on.

        Typically called by the owning module.

        Parameters
        ----------
        do_dx : Numpy 2D array Jacobian[i,j] of tree outputs[i] w.r.t. this input port elements[j].

        Returns
        -------
        ready : List of modules to backprop next.
        """
        if self._source is not None:
            return self._source.backprop(do_dx)
        else:
            return []

    def register_source(self, src):
        """Register an OutputPort source for this port.

        Parameters
        ----------
        src : OutputPort to take as the source of this port.
        """
        if not isinstance(src, OutputPort):
            raise ValueError('src must be OutputPort')
        self._source = src

    @property
    def value(self):
        return self._value

class OutputPort(object):
    """An output from a module. Typically instantiated as a module member.

    Parameters
    ----------
    module : The owning module. Must implement the ModuleBase interface.
    """
    def __init__(self, module):
        if not isinstance(module, ModuleBase):
            raise ValueError('module must implement ModuleBase')

        self._module = module

        self._backprop_acc = None
        self._num_backs = 0

        self._value = None
        self._consumers = []

    def __nonzero__(self):
        """Override of Python boolean test operator to return whether this port has a value.
        """
        return self.value is not None

    @property
    def num_consumers(self):
        """Return the number of registered consumers.
        """
        return len(self._consumers)

    @property
    def value(self):
        return self._value

    def register_consumer(self, con):
        """Register an InputPort consumer to this port.
        """
        if not isinstance(con, InputPort):
            raise ValueError('Consumer must be InputPort')
        self._consumers.append(con)

    def invalidate(self):
        """Invalidate this port and propagate.

        Returns
        -------
        valid : List of modules to invalidate next
        """
        # If we're already invalid, there's nothing to do
        if not self:
            return []

        self._backprop_acc = None
        self._num_backs = 0
        self._value = None

        valid = []
        if not self._module.is_invalid():
            valid.append(self._module)

        for con in self._consumers:
            valid += con.invalidate()
        return valid

    def foreprop(self, v):
        """Perform forward-propagation through this output.

        Typically called by the owning module.

        Parameters
        ----------
        v : The value to set this port to.

        Returns
        -------
        ready : List of modules to foreprop next.
        """
        self._value = v
        ready = []
        for con in self._consumers:
            ready += con.foreprop(self._value)
        return ready

    def backprop(self, do_dx):
        """Perform backward-propagation through this output.

        Typically called by a connected InputPort.
        Only propagates when data from all registered consumers is received.

        Parameters
        ----------
        do_dx : Numpy 2D array Jacobian[i,j] of tree outputs[i] w.r.t. this input port elements[j]

        Returns
        -------
        ready : List of modules to backprop next
        """
        if do_dx is None:
            raise RuntimeError('OutputPort received None backprop value.')

        do_dx.tick_descent()
        if self._backprop_acc is None:
            self._backprop_acc = do_dx
        else:
            self._backprop_acc += do_dx
        self._num_backs += 1

        # Check for backprop errors
        if self._num_backs > len(self._consumers):
            errstr = 'Received %d backprops for %d consumers!' % (self._num_backs, len(self._consumers))
            raise RuntimeError(errstr)

        # If we've heard from every consumer and our module is ready
        if self.backprop_ready() and self._module.backprop_ready():
            return [self._module]
        else:
            return []

    def backprop_ready(self):
        """Returns if this port has heard from all its consumers.
        """
        return self._num_backs == self.num_consumers

    def chain_backprop(self, dy_dx=None):
        """Returns a copy of this port's backprop accumulator right-multiplied by the
        given gradient. If the port has not received a backprop, returns None.
        """
        if self._backprop_acc is None:
            return None
            #raise RuntimeError('Cannot chain backprop! Port has not received do_dx.')

        out_acc = self._backprop_acc.copy()
        if dy_dx is not None:
            out_acc = out_acc * dy_dx
        return out_acc

    @property
    def backprop_accumulator(self):
        """Returns the port's backprop accumulator.
        """
        return self._backprop_acc

    @property
    def backprop_value(self):
        if self._backprop_acc is None:
            return 0
        else:
            return self._backprop_acc.retrieve()

def link_ports(in_port, out_port):
    """Join an input and output port together.

    Parameters
    ----------
    in_port  : InputPort to join
    out_port : OutputPort to join
    """
    if not isinstance(in_port, InputPort):
        raise ValueError('in_port must be an InputPort.')
    if not isinstance(out_port, OutputPort):
        raise ValueError('out_port must be an OutputPort.')

    in_port.register_source(out_port)
    out_port.register_consumer(in_port)

# @profile
def iterative_operation(init_module, op):
    # TODO Allow taking list of initial modules
    """Iteratively perform an operation on a module tree.

    This function should be used instead of recursive calls, which do not scale
    to deep trees very well.

    Parameters
    ----------
    init_module : Module to begin iteration on
    op          : Function that takes a module and returns a list of modules to operate on next
    """
    to_prop = deque()
    to_prop.append(init_module)
    while len(to_prop) > 0:
        current = to_prop.popleft()
        ready_children = op(current)
        for c in ready_children:
            to_prop.append(c)

def iterative_foreprop(init_module):
    """Iterative forward-pass propagation on a module tree.
    """
    op = lambda x: x.foreprop()
    iterative_operation(init_module, op)

def iterative_backprop(init_module):
    """Iterative backward-pass propagation on a module tree.
    """
    op = lambda x: x.backprop()
    iterative_operation(init_module, op)

def iterative_invalidate(init_module):
    """Iterative invalidation on a module tree.
    """
    op = lambda x: x.invalidate()
    iterative_operation(init_module, op)

File: src/ModulesCore.cpp
#include "modprop/compo/ModulesCore.h"
#include <iostream>
#include <algorithm>

namespace argus
{
ModuleBase::ModuleBase() {}

ModuleBase::~ModuleBase()
{}

void ModuleBase::RegisterInput( InputPort * in )
{
	_inputs.push_back( in );
}

void ModuleBase::RegisterOutput( OutputPort * out )
{
	_outputs.push_back( out );
}

void ModuleBase::UnregisterInput( InputPort * in )
{
	std::vector<InputPort *>::iterator iter;
	iter = std::find( _inputs.begin(), _inputs.end(), in );
	if( iter == _inputs.end() )
	{
		throw std::runtime_error( "Cannot unregister non-registered input port" );
	}
	_inputs.erase( iter );
}

void ModuleBase::UnregisterOutput( OutputPort * out )
{
	std::vector<OutputPort *>::iterator iter;
	iter = std::find( _outputs.begin(), _outputs.end(), out );
	if( iter == _outputs.end() )
	{
		throw std::runtime_error( "Cannot unregister non-registered output port" );
	}
	_outputs.erase( iter );
}

void ModuleBase::UnregisterAllSources( bool recurse )
{
	BOOST_FOREACH ( InputPort * in, _inputs )
	{
		in->UnregisterSource( recurse );
	}
}

void ModuleBase::UnregisterAllConsumers( bool recurse )
{
	BOOST_FOREACH ( OutputPort * out, _outputs )
	{
		out->UnregisterAllConsumers( recurse );
	}
}

bool ModuleBase::FullyValid() const
{
	BOOST_FOREACH ( InputPort * in, _inputs )
	{
		if( !in->Valid() )
		{
			return false;
		}
	}
	return true;
}

bool ModuleBase::FullyInvalid() const
{
	BOOST_FOREACH ( InputPort * in, _inputs )
	{
		if( in->Valid() )
		{
			return false;
		}
	}
	BOOST_FOREACH ( OutputPort * out, _outputs )
	{
		if( out->Valid() )
		{
			return false;
		}
	}
	return true;
}

bool ModuleBase::BackpropReady()
{
	BOOST_FOREACH ( OutputPort * out, _outputs )
	{
		if( !out->BackpropReady() )
		{
			return false;
		}
	}
	return true;
}

void ModuleBase::Invalidate()
{
	BOOST_FOREACH ( InputPort * in, _inputs )
	{
		in->Invalidate();
	}
	BOOST_FOREACH ( OutputPort * out, _outputs )
	{
		out->Invalidate();
	}
}

std::string ModuleBase::ToString() const
{
	std::stringstream ss;
	ss << "Module:" << std::endl;
	BOOST_FOREACH( InputPort * in, _inputs )
	{
		ss << *in << std::endl;
	}
	BOOST_FOREACH( OutputPort * out, _outputs )
	{
		ss << *out << std::endl;
	}
	return ss.str();
}

std::ostream& operator<<( std::ostream& os, const ModuleBase& mod )
{
	os << mod.ToString();
	return os;
}

InputPort::InputPort( ModuleBase& base )
	: _module( base ), _valid( false ), _source( nullptr )
{}

InputPort::~InputPort()
{
	// UnregisterSource( true );
}

void InputPort::RegisterSource( OutputPort * src )
{
	_source = src;
}

void InputPort::UnregisterSource( bool recurse )
{
	if( _source && recurse )
	{
		_source->UnregisterConsumer( this, false );
	}
	_source = nullptr;
}

bool InputPort::Valid() const
{
	return _valid;
}

void InputPort::Invalidate()
{
	if( !Valid() )
	{
		return;
	}

	_value = MatrixType();
	_valid = false;

	// if( !_module.FullyInvalid() )
	// {
	_module.Invalidate();
	// }
	if( _source ) //&& _source->Valid() )
	{
		_source->Invalidate();
	}
}

void InputPort::Foreprop( const MatrixType& val )
{
	if( _valid )
	{
		throw std::runtime_error( "Already valid input received foreprop" );
	}

	_valid = true;
	_value = val;

	if( _module.FullyValid() )
	{
		_module.Foreprop();
	}
}

void InputPort::Backprop( const MatrixType& dodx )
{
	if( _source )
	{
		_source->Backprop( dodx );
	}
}

const MatrixType& InputPort::GetValue() const
{
	if( !_valid )
	{
		throw std::runtime_error( "Cannot get value from invalid input port" );
	}
	return _value;
}

std::string InputPort::ToString() const
{
	std::stringstream ss;
	ss << "Input port:" << std::endl
	<< "\tSource: " << _source << std::endl
	<< "\tValid: " << _valid << std::endl;
	return ss.str();
}

std::ostream& operator<<( std::ostream& os, const InputPort& in )
{
	os << in.ToString();
	return os;
}

OutputPort::OutputPort( ModuleBase& base )
	: _module( base ), _valid( false ), _numBacks( 0 )
{}

OutputPort::~OutputPort()
{
	// UnregisterAllConsumers( true );
}

bool OutputPort::Valid() const
{
	return _valid;
}

size_t OutputPort::NumConsumers() const
{
	return _consumers.size();
}

void OutputPort::RegisterConsumer( InputPort * in )
{
	if( !in )
	{
		throw std::runtime_error( "Cannot register null consumer!" );
	}
	_consumers.push_back( in );
}

void OutputPort::UnregisterConsumer( InputPort * in, bool recurse )
{
	std::vector<InputPort *>::iterator iter;
	iter = std::find( _consumers.begin(), _consumers.end(), in );
	if( iter == _consumers.end() )
	{
		throw std::invalid_argument( "Cannot unregister non-registered consumer" );
	}
	if( recurse )
	{
		(*iter)->UnregisterSource( false );
	}
	_consumers.erase( iter );
}

void OutputPort::UnregisterAllConsumers( bool recurse )
{
	std::vector<InputPort *> conCopy = _consumers;
	BOOST_FOREACH ( InputPort * con, conCopy )
	{
		UnregisterConsumer( con, recurse );
	}
}

void OutputPort::Invalidate()
{
	if( !Valid() )
	{
		return;
	}

	_backpropAcc = MatrixType();
	_numBacks = 0;
	_value = MatrixType();
	_valid = false;

	// if( !_module.FullyInvalid() )
	// {
	_module.Invalidate();
	// }
	BOOST_FOREACH ( InputPort * con, _consumers )
	{
		// if( con->Valid() )
		// {
		con->Invalidate();
		// }
	}
}

void OutputPort::Foreprop( const MatrixType& val )
{
	_value = val;
	_valid = true;
	BOOST_FOREACH ( InputPort * con, _consumers )
	{
		con->Foreprop( _value );
	}
}

void OutputPort::Backprop( const MatrixType& dodx )
{
	if( !_valid )
	{
		throw std::runtime_error( "Cannot backprop invalid output port!" );
	}

	if( dodx.cols() != _value.size() )
	{
		std::cout << "dodx: " << dodx << std::endl;
		std::cout << "value: " << _value << std::endl;
		throw std::runtime_error( "Output backprop dimension mismatch" );
	}

	if( dodx.size() == 0 )
	{
		throw std::runtime_error( "Received empty backprop value" );
	}

	if( _backpropAcc.size() == 0 )
	{
		_backpropAcc = dodx;
	}
	else
	{
		if( _backpropAcc.rows() != dodx.rows() ||
		    _backpropAcc.cols() != dodx.cols() )
		{
			throw std::runtime_error( "Backprop value size mismatch" );
		}
		_backpropAcc += dodx;
	}

	_numBacks++;

	if( _numBacks > _consumers.size() )
	{
		throw std::runtime_error( "Received more backprops than consumers!" );
	}

	if( BackpropReady() && _module.BackpropReady() )
	{
		_module.Backprop();
	}
}

bool OutputPort::BackpropReady() const
{
	return _numBacks == _consumers.size();
}

MatrixType OutputPort::ChainBackprop( const MatrixType& dydx )
{
	if( _backpropAcc.size() == 0 )
	{
		return MatrixType();
	}

	if( dydx.size() != 0 )
	{
		return _backpropAcc * dydx;
	}
	else
	{
		return _backpropAcc;
	}
}

const MatrixType& OutputPort::GetBackpropValue() const
{
	return _backpropAcc;
}

const MatrixType& OutputPort::GetValue() const
{
	if( !_valid )
	{
		throw std::runtime_error( "Cannot get value from invalid output port" );
	}
	return _value;
}

std::string OutputPort::ToString() const
{
	std::stringstream ss;
	ss << "Output port:" << std::endl
	<< "\tNum consumers: " << NumConsumers() << std::endl
	<< "\tValid: " << _valid << std::endl
	<< "\tNum backprops: " << _numBacks;
	return ss.str();
}

std::ostream& operator<<( std::ostream& os, const OutputPort& out )
{
	os << out.ToString();
	return os;
}

void link_ports( OutputPort& out, InputPort& in )
{
	in.RegisterSource( &out );
	out.RegisterConsumer( &in );
}

void unlink_ports( OutputPort& out, InputPort& in )
{
	in.UnregisterSource( false );
	out.UnregisterConsumer( &in, false );
}

MatrixType sum_matrices( const std::vector<MatrixType>& mats )
{
	if( mats.size() == 0 )
	{
		throw std::runtime_error( "Cannot sum empty vector of matrices" );
	}

	bool init = false;
	MatrixType out;
	for( unsigned int i = 0; i < mats.size(); ++i )
	{
		if( mats[i].size() == 0 )
		{
			continue;
		}
		if( !init )
		{
			out = mats[i];
			init = true;
			continue;
		}
		out += mats[i];
	}

	if( !init )
	{
		throw std::runtime_error( "Received no non-empty matrices" );
	}
	return out;
}
}